### PR TITLE
Add flags to VarInfo

### PR DESCRIPTION
**Pull request by xukai92**
Resolving #389 and #391 by adding another filed to `VarInfo` called `flags`

TODOs
- [x] get rid of the use of `NULL`
- [x] turn `flags` to `VarName`-wise
  - Sparse by dictionary or not?
- [x] genearlize the signature of flags by ~~macro~~ better function signature

### CODE REVIEW

**Comment by xukai92** [comment]
@yebai Do we still want the feature of VarInfo which supports storing multiple chains in the same time, which we never use, i.e. make vi.spl a vector of vector.

**Comment by xukai92** [comment]
@emilemathieu I introduced a function `set_retained_vns_del_by_spl!` in `varinfo.jl` to replace our previous `getretain` and set to `NULL` statement (the function is almost the same as `getretain`). Because of it we actually don't need `getretain`. Do think we can just remove it?

**Comment by xukai92** [comment]
Another design issues I was thinking of is: should we make `flags` sparse or dense

i) sparse: if we assume only a small part of variables will be set a flag, a set/hash map to represent the flag is more memory-efficient
ii) dense: if most of the variables will be set flags, an array based implement is good

File: src/core/varinfo.jl [hunk 1]
```diff
@@ -310,8 +318,23 @@ end
 # Rand & replaying method for VarInfo #
 #######################################
 
-# Check if a vn is set to NULL
-isnan(vi::VarInfo, vn::VarName) = any(isnan.(getval(vi, vn)))
+check_flag(vi::VarInfo, vn::VarName, flag::String) = vi.flags[flag][getidx(vi, vn)]
```
Comment (emilemathieu) [inline src/core/varinfo.jl]: `get_flag` instead ?

File: src/core/varinfo.jl [hunk 2]
```diff
@@ -310,8 +318,23 @@ end
...
+
+set_retained_vns_del_by_spl!(vi::VarInfo, spl::Sampler) = begin
+  gidcs = getidcs(vi, spl)
+  if vi.num_produce == 0
+    for i = length(gidcs):-1:1
+      vi.flags["del"][gidcs[i]] = true
+    end
+  else
+    retained = [idx for idx in 1:length(vi.orders) if idx in gidcs && vi.orders[idx] > vi.num_produce]
+    for i = retained
```
Comment (emilemathieu) [inline src/core/varinfo.jl]: `i = 1:retained` ?

File: src/samplers/pgibbs.jl [hunk 3]
```diff
@@ -56,8 +56,8 @@ step(model::Function, spl::Sampler{PG}, vi::VarInfo) = begin
   ref_particle = isempty(vi) ?
                  nothing :
                  forkr(Trace(model, spl, vi))
-
-  vi[getretain(vi, spl)] = NULL
+
```
Comment (emilemathieu) [inline src/samplers/pgibbs.jl]: empty space

**Comment by emilemathieu** [comment]
Maybe a test could be useful (or modifying existing test)s to check that `vi.flags` is built and updated as planned. Otherwise, this PR seems right !

File: src/core/varinfo.jl (hunk 1 above)
Comment (xukai92) [reply src/core/varinfo.jl]: Or maybe `is_flagged`? I was trying to find a name which fits the use case of this function. As it always return a `Bool` type, we always use it like 
```julia
if is_flagged(vi, vn, "missing")
  ...
end
```

File: src/core/varinfo.jl (hunk 2 above)
Comment (xukai92) [reply src/core/varinfo.jl]: `retained` here is a vector of indices, and `i = retained` is equivalent to `i in retained`. But probably `i in retained` is more readable.

**Comment by xukai92** [comment]
Good suggestion. I will add some tests for `vi.flags`!

**Comment by xukai92** [comment]
Left to discuss
- [ ] merge trans or not
- [ ] sparse or dense

**Comment by yebai** [comment]
@xukai92 Perhaps we can solve the following 3 tasks and merge this PR first?

- [ ] merge trans ~~or not~~
- [x] ~~sparse or~~ dense
- [ ] remove nested Array

We can turn ·VarInfo` into a submodule in a seprate PR.

**Comment by xukai92** [comment]
Sure. will do it before Wednesday

**Comment by xukai92** [comment]
@yebai Can you review the PR plz? I tried to modualize VarInfo but meet some issues - will probably do that in another PR.

**Comment by xukai92** [comment]
Oops. Seems my modualization works now. It's a bit ugly as I literature export all functions required by Turing.jl from VarRepaly - but I guess we can solve the namespace issue later (?)

File: src/samplers/pgibbs.jl [hunk 4]
```diff
@@ -56,8 +56,8 @@ step(model::Function, spl::Sampler{PG}, vi::VarInfo) = begin
   ref_particle = isempty(vi) ?
                  nothing :
                  forkr(Trace(model, spl, vi))
-
-  vi[getretain(vi, spl)] = NULL
+  
+  set_retained_vns_del_by_spl!(vi, spl)
```
Comment (yebai) [inline src/samplers/pgibbs.jl]: I think the previous API is more readable, I would prefer something like 

```
vi.flags[spl] = DELETE #similiarly vi.flags[spl] = TRANSFORMED
```

**Pull request merged**